worker/formatting_stack/kondo_classpath_cache.clj

```diff
@@ -4,22 +4,13 @@
   Only needs to be created once, as the classpath never changes."
   (:require
    [clj-kondo.core :as kondo]
-   [clojure.string :as string])
-  (:import
-   (java.io File)))
-
-;; Use kondo's official default config dir, so that we don't bloat consumers' project layouts:
-(def cache-parent-dir ".clj-kondo")
-
-;; Don't use .clj-kondo directly since it can be accessed concurrently (e.g. f-s + a second Kondo from VS Code):
-(def cache-subdir "formatting-stack-cache")
-
-(def cache-dir (str cache-parent-dir File/separator cache-subdir))
+   [clojure.string :as string]))
 
 (def classpath-cache
   (future
     (let [files (-> (System/getProperty "java.class.path")
                     (string/split #"\:"))]
-      (-> (File. cache-parent-dir cache-subdir) .mkdirs)
-      (kondo/run! {:lint      files
-                   :cache-dir cache-dir}))))
+      (->> (partition-all 100 files) ;; prevent OOM when classpath is large
```

Comment:
Nice find here btw. You might want to report it upstream - they can plausibly value it, and preferrably we wouldn't have to use APIs in a particularly careful manner.

I'd also be a bit afraid of the runtime impact of making N consecutive `kondo/run!` invocations (i.e. it's possibly slower). My biggest `(-> (System/getProperty "java.class.path") (string/split #"\:") count)` has 800+ items in it, implying 9 run! invocations.

Reply:
I reported this in Clojurians yesterday, but no responses unfortunately. Ill create an issue soon to see what the maintainer thinks.

---


The 9 invocations seem like a lot, but I don't think it's more significant than 1 invocation at 9times the size.

In my limited testing I could only compare runtime between creating the cache using `clj-kondo` command directly and the partitioned runs (because OOM). The wall-to-wall times were pretty comparable (~40s in my case). Lmk your experiences though!

Reply:
I responded over Clojurians with a possibly useful insight.

Perhaps you can repro this deterministically by intentionally setting  a low Xmx (say, `512m`) and an intentionally large input?

Maybe clj-kondo could ensure that its `:lint` input is processed in a lazy manner - that way, it would fit in memory at all times.

Reply:
> The 9 invocations seem like a lot, but I don't think it's more significant than 1 invocation at 9times the size.

Would have to measure. I'd also be afraid of constructing  an analysis cache without having the "whole picture" (the entire classpath) at hand. Otherwise it's well plausible that analysis will have a partial view and clj-kondo will be unable to accurately resolve namespaces, vars, etc.